### PR TITLE
Fix broken ingress

### PR DESCRIPTION
**Pull request by costinm**
Ingress was not working at all - this change makes it work enough to get certmanager to get certs, and will likely work for most common uses of Ingress.

What is not supported ( and didn't work before either ) is the TLS part: users should manage certificates using the Gateway config and associated secretes in ingressgateway. Since getting a cert and configuring the secret loading are complex operations - setting a proper Gateway is not a huge extra step, while loading them from Ingress resources would be confusing and complex.

With the new code:
- ingresses are using proper 'host' in the generated VirtualService, instead of a match on authority
- all ingresses with same host are merged - so it looks more like a VirtualService
- the host=*, path=/.* (default) in ingress is no longer supported - it would take over the entire gateway.

To allow use cases like certmanager, the generated Ingress is merged with user-specified VirtualService with same host by appending Ingress-defined rules at the end. Probably not perfect - but should work
well enough while users migrate away from Ingress and to support other apps creating Ingress using k8s APIs. 

Note that since no gateway is generated, users will need to great a Gateway with the proper name and select their desired gateway in order for ingress to be active. This can be added to the current ingress deployment or it can point to ingressgateway, both seem to work. We can discuss if we still want to keep the old ingress load balancer/IP/workload.

### CODE REVIEW

File: pilot/pkg/config/kube/ingress/controller.go [hunk 1]
```diff
@@ -185,11 +193,16 @@ func (c *controller) List(typ, namespace string) ([]model.Config, error) {
...
 		case model.VirtualService.Type:
-			_, virtualServices := ConvertIngressV1alpha3(*ingress, c.domainSuffix)
-			out = append(out, virtualServices)
+			ConvertIngressVirtualService(*ingress, c.domainSuffix, ingressByHost)
 		case model.Gateway.Type:
-			gateways, _ := ConvertIngressV1alpha3(*ingress, c.domainSuffix)
-			out = append(out, gateways)
+			//gateways := ConvertIngressV1alpha3(*ingress, c.domainSuffix)
+			//out = append(out, gateways)
+		}
```
Comment (rshriram) [inline pilot/pkg/config/kube/ingress/controller.go]: The auto generated gateway was a big deal for the ingress. As such our conversion tool is sufficient to convert to gateway and virtual service.

File: pilot/pkg/config/kube/ingress/conversion_test.go [hunk 2]
```diff
@@ -21,10 +21,123 @@ import (
...
+	ConvertIngressVirtualService(ingress, "mydomain", cfgs)
+	ConvertIngressVirtualService(ingress2, "mydomain", cfgs)
+
+	if len(cfgs) != 3 {
+		t.Error("VirtualServices, expected 3 got ", len(cfgs))
+	}
+
+	for n, cfg := range cfgs {
+		// Not clear if this is right - should probably be under input ns
+		if cfg.ConfigMeta.Namespace != "istio-system" {
```
Comment (rshriram) [inline pilot/pkg/config/kube/ingress/conversion_test.go]: You cannot have an ingress controller per namespace. So it doesn’t matter which namespace we convert all the ingress resource into.

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 3]
```diff
@@ -405,6 +408,48 @@ func (configgen *ConfigGeneratorImpl) buildGatewayInboundHTTPRouteConfig(
...
+// [host]/path->backend configuration
+func mergeIngress(configs []model.Config) []model.Config {
+	configsOut := []model.Config{}
+	hostsToVS := map[string]*networking.VirtualService{}
+
+	// Ignore ingress resources, build a map of hosts to (real) VirtualServices.
+	for _, c := range configs {
+		vs := c.Spec.(*networking.VirtualService)
+		for _, h := range vs.Hosts {
+			if !strings.HasSuffix(c.Name, model.IstioIngressGatewayName) {
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/gateway.go]: This is deduplicating for non ingress. Why?

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 4]
```diff
@@ -405,6 +408,48 @@ func (configgen *ConfigGeneratorImpl) buildGatewayInboundHTTPRouteConfig(
...
+	}
+
+	for _, c := range configs {
+		vs := c.Spec.(*networking.VirtualService)
+		for _, h := range vs.Hosts {
+			if strings.HasSuffix(c.Name, model.IstioIngressGatewayName) {
+				vse, e := hostsToVS[h]
+				if e {
+					vse.Http = append(vse.Http, vs.Http...)
+					log.Infof("Merging ingress to vs %s %s", h, c.Name)
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/gateway.go]: The append was already done in the conversion. Why iterate over this again?

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 5]
```diff
@@ -405,6 +408,48 @@ func (configgen *ConfigGeneratorImpl) buildGatewayInboundHTTPRouteConfig(
...
+	hostsToVS := map[string]*networking.VirtualService{}
+
+	// Ignore ingress resources, build a map of hosts to (real) VirtualServices.
+	for _, c := range configs {
+		vs := c.Spec.(*networking.VirtualService)
+		for _, h := range vs.Hosts {
+			if !strings.HasSuffix(c.Name, model.IstioIngressGatewayName) {
+				vse, e := hostsToVS[h]
+				if e {
+					log.Warnf("Duplicated hostname %s OLD:%v NEW:%v", h, vse, vs)
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/gateway.go]: ?? Why are you doing this for non ingress stuff?

File: pilot/pkg/config/kube/ingress/controller.go (hunk 1 above)
Comment (costinm) [reply pilot/pkg/config/kube/ingress/controller.go]: I can uncomment it - but it doesn't really work. Secrets from TLS can't be loaded (there is no code to read the secrets in ingress - they need to be mounted). Other than that - a host-* and port 80/443 is all gateway needed, why generate one gateway per ingress ? 

I will add the Gateway with the proper name to helm - so users will not have to do anything, but separate PR since there is a whole separate discussion on how to setup secrets for TLS and which LB to use.

File: pilot/pkg/config/kube/ingress/conversion_test.go (hunk 2 above)
Comment (costinm) [reply pilot/pkg/config/kube/ingress/conversion_test.go]: Ack - yes, the secrets must be in same namespace with the gateway.

File: pilot/pkg/networking/core/v1alpha3/gateway.go (hunk 3 above)
Comment (costinm) [reply pilot/pkg/networking/core/v1alpha3/gateway.go]: Sorry, mixed separate fix - it seems we still generate duplicate domains and RDS rejects it. 
I can send you the log.

I can remove this part of the PR and move it to separate.

File: pilot/pkg/networking/core/v1alpha3/gateway.go (hunk 4 above)
Comment (costinm) [reply pilot/pkg/networking/core/v1alpha3/gateway.go]: Good point, will remove ( I implemented this first - then figured it's easy to merge during conversion ).

File: pilot/pkg/networking/core/v1alpha3/gateway.go (hunk 5 above)
Comment (costinm) [reply pilot/pkg/networking/core/v1alpha3/gateway.go]: See above, will move to separate PR or file a bug. We probably need to do same trick Nate used to use the timestamp to reject most recent virtual service in cases of duplicated hostname.

File: pilot/pkg/config/kube/ingress/controller.go [hunk 6]
```diff
@@ -112,15 +112,20 @@ func (c *controller) RegisterEventHandler(typ string, f func(model.Config, model
...
+		// the inputs.
+		// In future we may do smarter processing - but first we'll do
+		// major refactoring. No need to recompute everything and generate
+		// multiple events.
+
 		// TODO: This works well for Add and Delete events, but not so for Update:
 		// An updated ingress may also trigger an Add or Delete for one of its constituent sub-rules.
 		switch typ {
 		case model.Gateway.Type:
-			config, _ := ConvertIngressV1alpha3(*ingress, c.domainSuffix)
```
Comment (rshriram) [inline pilot/pkg/config/kube/ingress/controller.go]: nit. If you are going to be adding a gateway for the ingress, you can essentially get rid of the whole Gateway case and have just the model.VirtualService ..

File: pilot/pkg/config/kube/ingress/controller.go [hunk 7]
```diff
@@ -185,14 +193,19 @@ func (c *controller) List(typ, namespace string) ([]model.Config, error) {
 
 		switch typ {
 		case model.VirtualService.Type:
-			_, virtualServices := ConvertIngressV1alpha3(*ingress, c.domainSuffix)
-			out = append(out, virtualServices)
+			ConvertIngressVirtualService(*ingress, c.domainSuffix, ingressByHost)
 		case model.Gateway.Type:
-			gateways, _ := ConvertIngressV1alpha3(*ingress, c.domainSuffix)
+			gateways := ConvertIngressV1alpha3(*ingress, c.domainSuffix)
```
Comment (rshriram) [inline pilot/pkg/config/kube/ingress/controller.go]: same comment as above. and you can get rid fo the convertIngressV1alpha3 as well

File: pilot/pkg/config/kube/ingress/conversion.go [hunk 8]
```diff
@@ -105,22 +106,49 @@ func ConvertIngressV1alpha3(ingress v1beta1.Ingress, domainSuffix string) (model
...
+	gatewayConfig := model.Config{
+		ConfigMeta: model.ConfigMeta{
+			Type:      model.Gateway.Type,
+			Group:     model.Gateway.Group,
+			Version:   model.Gateway.Version,
+			Name:      ingress.Name + "-" + model.IstioIngressGatewayName,
+			Namespace: model.IstioIngressNamespace,
+			Domain:    domainSuffix,
+		},
+		Spec: gateway,
```
Comment (rshriram) [inline pilot/pkg/config/kube/ingress/conversion.go]: And when you add the gateway manually to helm, this entire function can go away.

File: pilot/pkg/config/kube/ingress/conversion_test.go (hunk 2 above)
Comment (rshriram) [reply pilot/pkg/config/kube/ingress/conversion_test.go]: There is also a constant for Istio-system FYI

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 9]
```diff
@@ -239,7 +240,8 @@ func (configgen *ConfigGeneratorImpl) createGatewayHTTPFilterChainOpts(
 		// Build a filter chain for each TLS server
 		for i, server := range servers {
 			rdsName := model.GatewayRDSRouteName(server)
-			routeCfg := configgen.buildGatewayInboundHTTPRouteConfig(env, node, nameToServiceMap, gatewayNames, []*networking.Server{server}, rdsName)
+			routeCfg := configgen.buildGatewayInboundHTTPRouteConfig(env, node, push,
+				nameToServiceMap, gatewayNames, []*networking.Server{server}, rdsName)
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/gateway.go]: Please get rid of this fully. I had kept this here while adding RDS but forgot to remove these things after RDS started working.

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 10]
```diff
@@ -239,7 +240,8 @@ func (configgen *ConfigGeneratorImpl) createGatewayHTTPFilterChainOpts(
...
 			if routeCfg == nil {
 				log.Warnf("omitting HTTP listeners for port %d filter chain %d due to no routes", server.Port, i)
 				continue
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/gateway.go]: Oh crap.. Also get rid of this if block. since `routeCfg` is no longer present.

File: pilot/pkg/config/kube/ingress/conversion.go [hunk 11]
```diff
@@ -105,22 +106,49 @@ func ConvertIngressV1alpha3(ingress v1beta1.Ingress, domainSuffix string) (model
...
 	}
 
-	var httpRoutes []*networking.HTTPRoute
+	return gatewayConfig
+}
+
+func ConvertIngressVirtualService(ingress v1beta1.Ingress, domainSuffix string, ingressByHost map[string]*model.Config) {
+	// Ingress allows a single host - if missing '*' is assumed
```
Comment (ostromart) [inline pilot/pkg/config/kube/ingress/conversion.go]: public functions should be documented. i suggest moving much of the text like "mixing *" to the function comment since it will explain the behavior of the function to the reader without needing to go into the code.

File: pilot/pkg/config/kube/ingress/conversion.go [hunk 12]
```diff
@@ -105,22 +106,49 @@ func ConvertIngressV1alpha3(ingress v1beta1.Ingress, domainSuffix string) (model
...
+	// host.
+
 	for _, rule := range ingress.Spec.Rules {
 		if rule.HTTP == nil {
 			log.Infof("invalid ingress rule %s:%s for host %q, no paths defined", ingress.Namespace, ingress.Name, rule.Host)
 			continue
 		}
 
+		host := rule.Host
+		namePrefix := strings.Replace(host, ".", "-", -1)
```
Comment (ostromart) [inline pilot/pkg/config/kube/ingress/conversion.go]: even though it's trivial, wrapping this in a helper function will document why this replace is needed.

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 13]
```diff
@@ -405,6 +420,68 @@ func (configgen *ConfigGeneratorImpl) buildGatewayInboundHTTPRouteConfig(
...
+						if len(r.Match) == 0 {
+							continue // unexpected
+						}
+						if p, ok := r.Match[0].Uri.GetMatchType().(*networking.StringMatch_Prefix); ok {
+							if p.Prefix == "/" || p.Prefix == "" {
+								log.Infof("Ignoring / prefix for %s %s", h, c.Name)
+								continue
+							}
+						}
+						vse.Http = append([]*networking.HTTPRoute{r}, vse.Http...)
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/gateway.go]: ??
This is merging user generated virtual services with user generated ingress. FYI, this is pretty much what we did before 0.8, and there was a never ending stream of bugs. Keeping the entire ingress separate made life much more easier. And we made a very conscious decision to stop with just the conversion and not do any runtime merging with user generated artifacts precisely due to all the bugs we faced. So, lets please not bring this back, one week before 1.0 launch :).

File: pilot/pkg/config/kube/ingress/conversion.go [hunk 14]
```diff
@@ -105,22 +106,49 @@ func ConvertIngressV1alpha3(ingress v1beta1.Ingress, domainSuffix string) (model
...
+			host = "*"
+		}
+		virtualService := &networking.VirtualService{
+			Hosts:    []string{},
+			Gateways: []string{model.IstioIngressGatewayName},
+		}
+
+		virtualService.Hosts = []string{host}
+
+		httpRoutes := []*networking.HTTPRoute{}
```
Comment (ostromart) [inline pilot/pkg/config/kube/ingress/conversion.go]: it's more efficient and idiomatic to write this as var httpRoutes []*networking.HTTPRoute, as there's no need to create an empty array at this point and better to leave it as nil.

File: pilot/pkg/networking/core/v1alpha3/gateway.go (hunk 13 above)
Comment (rshriram) [reply pilot/pkg/networking/core/v1alpha3/gateway.go]: Come to think of it, this entire logic is of no use. The gateway code is for a single gateway (istio: ingress, or istio: ingressgateway). So, the list of configs you have here for the current gateway being processed (istio: ingress) are all the auto generated gateways in conversion.go.

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 15]
```diff
@@ -337,32 +340,44 @@ func (configgen *ConfigGeneratorImpl) buildGatewayInboundHTTPRouteConfig(
...
-			}
-
-			if tlsRedirect[gatewayHost] {
-				host.RequireTls = route.VirtualHost_ALL
+			key := fmt.Sprintf("%s:%d", v.Name, port)
+			old, f := dedupHosts[key]
+			if f {
+				push.Add(model.DuplicatedDomains, key, node,
+					fmt.Sprintf("Duplicated virtual host %s use %s:%s rejecting %s:%s",
+						key, old.Namespace, old.Name, v.Namespace, v.Name))
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/gateway.go]: I don't think this is that simple (string match). Gateway's allow wildcards and we respect wildcards and absorb hosts under a given wildcard. The issues you were seeing were mostly due to the incorrect merging you attempt below. As such, you shouldn't be hitting any case of duplicate virtual hosts unless you generate two virtual services for saem ingress host, in conversion.go. So I would check the logic there.

File: pilot/pkg/config/kube/ingress/conversion.go [hunk 16]
```diff
@@ -131,40 +159,37 @@ func ConvertIngressV1alpha3(ingress v1beta1.Ingress, domainSuffix string) (model
...
-		ConfigMeta: model.ConfigMeta{
-			Type:      model.VirtualService.Type,
-			Group:     model.VirtualService.Group,
-			Version:   model.VirtualService.Version,
-			Name:      ingress.Name + "-" + model.IstioIngressGatewayName,
-			Namespace: model.IstioIngressNamespace,
-			Domain:    domainSuffix,
-		},
-		Spec: virtualService,
+		old, f := ingressByHost[host]
```
Comment (ostromart) [inline pilot/pkg/config/kube/ingress/conversion.go]: more idiomatic to call f ok. also, for ptrs, it's more conventional to test for zero value rather than using the bool ok value i.e. 
if old := ingressByHost[host]; old != nil {
see : https://blog.golang.org/go-maps-in-action

File: pilot/pkg/config/kube/ingress/conversion.go [hunk 17]
```diff
@@ -131,40 +159,37 @@ func ConvertIngressV1alpha3(ingress v1beta1.Ingress, domainSuffix string) (model
...
 	}
 
-	return gatewayConfig, virtualServiceConfig
+	// Matches * and "/". Currently not supported - would conflict
+	// with any other explicit VirtualService.
+	if ingress.Spec.Backend != nil {
+		log.Infof("Ignore default wildcard ingress, use VirtualService %s:%s",
+			ingress.Namespace, ingress.Name)
 
+	}
```
Comment (ostromart) [inline pilot/pkg/config/kube/ingress/conversion.go]: looking over this function, i don't see anything that would require ingress to be modified in place. per earlier comment, i'd recommend returning a new object instead.

File: pilot/pkg/config/kube/ingress/conversion.go [hunk 18]
```diff
@@ -233,6 +258,11 @@ func createStringMatch(s string) *networking.StringMatch {
 			MatchType: &networking.StringMatch_Prefix{Prefix: strings.TrimSuffix(s, ".*")},
 		}
 	}
+	if strings.HasSuffix(s, "/*") {
```
Comment (ostromart) [inline pilot/pkg/config/kube/ingress/conversion.go]: comment would be helpful here. the what is clear, but not really the why.

File: pilot/pkg/config/kube/ingress/conversion_test.go [hunk 19]
```diff
@@ -21,10 +21,123 @@ import (
 	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/intstr"
 
+	networking "istio.io/api/networking/v1alpha3"
+
```
Comment (ostromart) [inline pilot/pkg/config/kube/ingress/conversion_test.go]: remove blank line

File: pilot/pkg/config/kube/ingress/conversion_test.go [hunk 20]
```diff
@@ -21,10 +21,123 @@ import (
...
 
+func TestConversion(t *testing.T) {
+	ingress := v1beta1.Ingress{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Namespace: "mock", // goes into backend full name
+		},
+		Spec: v1beta1.IngressSpec{
+			Rules: []v1beta1.IngressRule{
+				{
+					Host: "my.host.com",
```
Comment (ostromart) [inline pilot/pkg/config/kube/ingress/conversion_test.go]: define this as const and use here and below.

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 21]
```diff
@@ -337,32 +340,44 @@ func (configgen *ConfigGeneratorImpl) buildGatewayInboundHTTPRouteConfig(
 
 	port := int(servers[0].Port.Number)
 	// NOTE: WE DO NOT SUPPORT two gateways on same workload binding to same virtual service
-	virtualServices := env.VirtualServices(gateways)
+	allVirtualServices := env.VirtualServices(gateways)
+	virtualServices := mergeIngress(allVirtualServices)
+
 	virtualHosts := make([]route.VirtualHost, 0, len(virtualServices))
+	dedupHosts := map[string]*model.Config{}
```
Comment (ostromart) [inline pilot/pkg/networking/core/v1alpha3/gateway.go]: use make

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 22]
```diff
@@ -405,6 +420,68 @@ func (configgen *ConfigGeneratorImpl) buildGatewayInboundHTTPRouteConfig(
...
+	configsOut := []*model.Config{}
+
+	hostsToVS := map[string]*networking.VirtualService{}
+
+	// Ignore ingress resources, build a map of hosts to (real) VirtualServices. VirtualService has priority.
+	for _, c := range configs {
+		vs := c.Spec.(*networking.VirtualService)
+		for _, h := range vs.Hosts {
+			if !strings.HasSuffix(c.Name, model.IstioIngressGatewayName) {
+				vse, e := hostsToVS[h]
```
Comment (ostromart) [inline pilot/pkg/networking/core/v1alpha3/gateway.go]: see earlier comment about testing for map keys.

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 23]
```diff
@@ -405,6 +420,68 @@ func (configgen *ConfigGeneratorImpl) buildGatewayInboundHTTPRouteConfig(
...
+	// Add the rules from Ingress generated resources, as well as
+	// all ingress generated resources for domains not covered by VirtualService
+
+	// We could restrict ingress merging if a VirtualService exists - only /.well-known (to narrow
+	// it to ACME case, or only exact match, etc.
+	for _, c := range configs {
+		vs := c.Spec.(*networking.VirtualService)
+		for _, h := range vs.Hosts {
+			if strings.HasSuffix(c.Name, model.IstioIngressGatewayName) {
+				vse, e := hostsToVS[h]
```
Comment (ostromart) [inline pilot/pkg/networking/core/v1alpha3/gateway.go]: same here

File: pilot/pkg/config/kube/ingress/controller.go [hunk 24]
```diff
@@ -185,14 +193,19 @@ func (c *controller) List(typ, namespace string) ([]model.Config, error) {
 
 		switch typ {
 		case model.VirtualService.Type:
-			_, virtualServices := ConvertIngressV1alpha3(*ingress, c.domainSuffix)
-			out = append(out, virtualServices)
+			ConvertIngressVirtualService(*ingress, c.domainSuffix, ingressByHost)
```
[the remark this answers is not in the document]
Comment (costinm) [reply pilot/pkg/config/kube/ingress/controller.go]: Why ? Seems to work fine.

File: pilot/pkg/config/kube/ingress/conversion.go (hunk 11 above)
Comment (costinm) [reply pilot/pkg/config/kube/ingress/conversion.go]: done.

File: pilot/pkg/config/kube/ingress/conversion.go (hunk 12 above)
Comment (costinm) [reply pilot/pkg/config/kube/ingress/conversion.go]: And likely make the code harder to read (context switch). I can add a comment if it is not obvious ( dot not valid in names)

File: pilot/pkg/config/kube/ingress/conversion_test.go (hunk 19 above)
Comment (costinm) [reply pilot/pkg/config/kube/ingress/conversion_test.go]: I'm running "make format" - I think manual management of imports (or spending time reviewing the order) is not best use of time, adding imports is also automated by the IDE.

File: pilot/pkg/config/kube/ingress/conversion_test.go (hunk 20 above)
Comment (costinm) [reply pilot/pkg/config/kube/ingress/conversion_test.go]: I think it also makes things harder to read. A string is immutable too.

File: pilot/pkg/config/kube/ingress/controller.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (ostromart) [reply pilot/pkg/config/kube/ingress/controller.go]: it's not clear from the function what is being mutated and how. also inconsistent with ConvertIngressV1alpha3 below.

File: pilot/pkg/config/kube/ingress/conversion.go (hunk 12 above)
Comment (ostromart) [reply pilot/pkg/config/kube/ingress/conversion.go]: why would namePrefix := replaceInvalidDotsInName(host) be harder to read? this is self commenting.

File: pilot/pkg/config/kube/ingress/conversion_test.go (hunk 19 above)
Comment (ostromart) [reply pilot/pkg/config/kube/ingress/conversion_test.go]: sure, i just happened to notice it.

File: pilot/pkg/model/gateway.go [hunk 25]
```diff
@@ -95,14 +95,18 @@ func MergeGateways(gateways ...Config) *MergedGateway {
...
 							continue
 						}
 						rdsRouteConfigNames[rdsName] = []*networking.Server{s}
 					}
 
 					// We have another TLS server on the same port. Can differentiate servers using SNI
+					if s.Tls == nil {
+						log.Warnf("TLS server without TLS options %s %s", name, s.String())
+						continue
+					}
```
Comment (rshriram) [inline pilot/pkg/model/gateway.go]: Please remove this. I have added validation to ensure that HTTPS/TLS servers should have TLS settings and plain text HTTP servers cannot have any TLS settings.

File: pilot/pkg/config/kube/ingress/conversion_test.go (hunk 20 above)
Comment (ostromart) [reply pilot/pkg/config/kube/ingress/conversion_test.go]: maybe, but it's prone to typos. compiler cannot check that you are using the right string. 
(actually myHostName doesn't seem less readable to me than "my.host.com"). 
anyway, it's not going to lead to a tragic error in this file so not a biggie, up to you.

File: pilot/pkg/model/gateway.go (hunk 25 above)
Comment (rshriram) [reply pilot/pkg/model/gateway.go]: https://github.com/istio/istio/pull/7175/files#diff-5a64dc2088020f89bc6d77354b9ebb22R414 .. Will ensure that you cannot have a TLS server with non TLS port, and vice versa.

File: pilot/pkg/config/kube/ingress/controller.go [hunk 26]
```diff
@@ -173,6 +178,9 @@ func (c *controller) List(typ, namespace string) ([]model.Config, error) {
 	}
 
 	out := make([]model.Config, 0)
+
+	ingressByHost := map[string]*model.Config{}
```
[the remark this answers is not in the document]
Comment (costinm) [reply pilot/pkg/config/kube/ingress/controller.go]: Sorry, I am a bit frustrated. The real problem with this code is that it allocates shitload of objects and is horribly inefficient. Would be great if reviews focused less on how maps are created and more on why they are created and how to structure the code to reuse. At this point any contributor who can figure out the spaghetti of structures and missdirections will have little problems with either of the styles of creating a map (or order of imports for that matter - most IDEs don't even expose that)

File: pilot/pkg/config/kube/ingress/controller.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (costinm) [reply pilot/pkg/config/kube/ingress/controller.go]: That's intended - the other method is supposed to be deleted soon, and all the passing by value and un-needed copy and allocations that is used in this (and lots of other places) will need to be fixed as well. When we know we have allocation/memory problems - keeping consistent with inefficient code is not really desirable.

File: pilot/pkg/config/kube/ingress/conversion.go (hunk 12 above)
Comment (costinm) [reply pilot/pkg/config/kube/ingress/conversion.go]: It is not - it doesn't say how they are replaced, why they are invalid ( they are not invalid in a hostname ), and in the unlikely event it will be reused accidentally can lead to bigger problems. It's a one liner. Maybe 'replaceAllDotsWithMinus' would be self commenting. In go it is considered ok to duplicate small pieces of code.

File: pilot/pkg/config/kube/ingress/conversion_test.go (hunk 19 above)
Comment (costinm) [reply pilot/pkg/config/kube/ingress/conversion_test.go]: If linter complains - it'll have to be fixed. Otherwise - it's far more valuable to spend your time looking at efficiency/correctness of the code and the big picture...

File: pilot/pkg/config/kube/ingress/conversion_test.go (hunk 20 above)
Comment (costinm) [reply pilot/pkg/config/kube/ingress/conversion_test.go]: I'm not writing code for the compiler :-) The proper way to do this test is to move it to a file and 
read the file - in which case there will be no constant. I wrote it this way only because I am not sure what future this code has and it may not be worth adding too many more tests for ingress if we'll get rid of it soon... 

The test was mostly to help debugging and fixing the code - actual coverage will be in the test env (separate PR updating the stability cluster - and someone will need to bring back ingress tests in simple e2e).

File: pilot/pkg/model/gateway.go (hunk 25 above)
Comment (costinm) [reply pilot/pkg/model/gateway.go]: Got it, thanks - will merge your PR and remove this. Forgot to check - do you warn with enough info or have a counter ? I found it very hard to use the errors in pilot without having the namespace and name of the config.

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 27]
```diff
@@ -259,9 +259,9 @@ func (configgen *ConfigGeneratorImpl) buildSidecarInboundListeners(env *model.En
 		}
 
 		listenerMapKey := fmt.Sprintf("%s:%d", endpoint.Address, endpoint.Port)
-		if _, exists := listenerMap[listenerMapKey]; exists {
+		if old, exists := listenerMap[listenerMapKey]; exists {
 			env.PushStatus.Add(model.ProxyStatusConflictInboundListener, node.ID, node,
-				fmt.Sprintf("Rejected %s for %s", instance.Service.Hostname, listenerMapKey))
+				fmt.Sprintf("Rejected %s, used %s for %s", instance.Service.Hostname, old.Service.Hostname, listenerMapKey))
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/listener.go]: I think this is outdated.. @nmittler revamped this logic. We now print all services that conflict. So you can undo the changes to this function

Comment (rshriram) [reply pilot/pkg/networking/core/v1alpha3/listener.go]: oh ignore.. this is for inbound
